### PR TITLE
Adding primitive types to improve drawing

### PR DESCRIPTION
**Pull request by CoryCharlton**
## Description
- Restructure Graphics.Core so folders correspond to namespace
- Added System.Drawing.Point and marked nanoFramework.UI.Point as obsolete
- Added System.Drawing.Size
- Moved Mathematics to Graphics.Core
- Added System.Drawing.Rectangle
- Added overloads to nanoFramework.UI.Bitmap to use rectangle/point
- Added System.Drawing.ContentAlignment

## Motivation and Context
This improves the experience of drawing to a Bitmap

## How Has This Been Tested?
- Unit tests
- On device

## Types of changes

- [X] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [X] New feature (non-breaking change which adds functionality to code)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [X] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [X] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [X] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [X] I have added new tests to cover my changes.

### CODE REVIEW

**Review by Ellerbach — APPROVED** [review]
Is there a way to bring couple of unit tests with the additions?
Also thanks for the comments for the commented code, that's perfect!
And thanks for the addition! There are few bindings using those Point, Size, Rectangle. So some migration to expect as well.

File: nanoFramework.Graphics.Core/nanoFramework/UI/Point.cs [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 // See LICENSE file in the project root for full license information.
 
+using System;
+
 namespace nanoFramework.UI
 {
     /// <summary>
     /// A point data structure to represent a position on a plane.
     /// </summary>
+    [Obsolete("Switch to System.Drawing.Point", false)]
```
Comment (Ellerbach) [inline nanoFramework.Graphics.Core/nanoFramework/UI/Point.cs:11]: What is the interest of keeping it if you've already adjusted the Bitmap to only take the new Point?

**Comment by CoryCharlton** [comment]
> Is there a way to bring couple of unit tests with the additions? Also thanks for the comments for the commented code, that's perfect! And thanks for the addition! There are few bindings using those Point, Size, Rectangle. So some migration to expect as well.

Sure, I'll add some unit tests today.

Any feedback/comments on my other review comments?

**Review by Ellerbach — APPROVED** [review]
Thanks a lot for the addition and the work here. It looks good all up. Kudos points for the tests, for the clarity of the comments.

**Comment by Ellerbach** [comment]
> Any feedback/comments on my other review comments?

Yes, few more things:

* Bump the version in both files
* This will require to update the native side as well. Can you make sure to do this and raise a PR as well? So we can merge both accordingly

**Comment by CoryCharlton** [comment]
@Ellerbach I updated the minor version in the version.json

Still getting my head around the native side of things. I assume you are asking for changes in https://github.com/nanoframework/nf-interpreter

Couple questions:

- Why do these changes require updates on the native side?
- What do I need to update on the native side? I see [this](https://github.com/nanoframework/nf-interpreter/blob/main/src/nanoFramework.Graphics/Graphics/Native/nanoFramework_Graphics_nanoFramework_UI_Bitmap.cpp) but it doesn't seem relevant for my changes.

**Comment by Ellerbach** [comment]
> Why do these changes require updates on the native side?

Because by changing the API, you are changing the signature. So you need to update this on the native side. The stub are generated for you, then it's a matter of adding those back in their place. So here:
* https://github.com/nanoframework/nf-interpreter/blob/main/src/nanoFramework.Graphics/Graphics/Native/nanoFramework_Graphics.h (you should adjust anything that has been changed, be careful, it's not just a copy/paste of the generated file)
* https://github.com/nanoframework/nf-interpreter/blob/main/src/nanoFramework.Graphics/Graphics/Native/nanoFramework_Graphics.cpp (here, you basically just copy/paste the generated file and as you'll the bottom signature is going to change)

And as all changes, you have to bump as well the versions!

And on the managed side (this PR), you also have to bump this file: https://github.com/nanoframework/nanoFramework.Graphics/blob/main/version.json with something like 1.2

Let me know if you have any question!

**Comment by josesimoes** [comment]
@CoryCharlton no way to change those NULLs on the declaration. That's part of the mechanism there and it has to happen.

**Comment by CoryCharlton** [comment]
> @CoryCharlton no way to change those NULLs on the declaration. That's part of the mechanism there and it has to happen.

Understood. I threw out a naive though regardint an alternative in #interpreter

**Comment by CoryCharlton** [comment]
Not sure why `nanoFramework / NuGets using latest version` and `nanoFramework / package.lock exists` workflows aren't completing but I believe this should have everything needed. Please let me know if there is anything else required.

**Comment by josesimoes** [comment]
> Not sure why `nanoFramework / NuGets using latest version` and `nanoFramework / package.lock exists` workflows aren't completing 

They need permission to run when originating from external developer.

File: nanoFramework.Graphics.Core/System/Drawing/Point.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
```
Comment (josesimoes) [inline nanoFramework.Graphics.Core/System/Drawing/Point.cs:4]: Not 100% if this is safe to use... our execution engine may not be able to honor this request...

File: nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs [hunk 3]
```diff
@@ -0,0 +1,354 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
```
Comment (josesimoes) [inline nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs:4]: same here

File: nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+using nanoFramework.UI;
+
+namespace System.Drawing
+{
+    /// <summary>
+    /// Stores the location and size of a rectangular region.
+    /// </summary>
+    [Serializable]
+    public struct Rectangle // : IEquatable<Rectangle>
```
Comment (josesimoes) [inline nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs:13]: If `IEquatable`` is not needed better remove.

File: nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    {
+        /// <summary>
+        /// Creates a new instance of the <see cref='Rectangle'/> class with member data left uninitialized.
+        /// </summary>
+        public static readonly Rectangle Empty = new();
+
+        private int x; // Do not rename (binary serialization)
```
Comment (josesimoes) [inline nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs:20]: Per project code style: comments above the line not "inlined" please

File: nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+        }
+
+        /// <summary>
+        /// Creates a rectangle that represents the intersection between a and b. If there is no intersection, an
+        /// empty rectangle is returned.
+        /// </summary>
+        public static Rectangle Intersect(Rectangle a, Rectangle b)
+        {
+            // TODO: The documentation for Graphics says the Math library is required but it's not referenced at this time
+            int x1 = Mathematics.Max(a.X, b.X);// Math.Max(a.X, b.X);
```
Comment (josesimoes) [inline nanoFramework.Graphics.Core/System/Drawing/Rectangle.cs:298]: Min and Max are available in mscorlib as `private` for situations like this one.
Just need to add nanoFramework.Graphics to "Friends" list.

File: nanoFramework.Graphics.Core/System/Drawing/Size.cs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    /// </summary>
+    [Serializable]
+    public struct Size // : IEquatable<Size>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref='Size'/> class.
+        /// </summary>
+        public static readonly Size Empty;
+
+        private int width; // Do not rename (binary serialization)
```
Comment (josesimoes) [inline nanoFramework.Graphics.Core/System/Drawing/Size.cs:18]: Comment position!

File: nanoFramework.Graphics.Core/nanoFramework/UI/Mathematics.cs [hunk 8]
```diff
@@ -4,6 +4,7 @@
 // See LICENSE file in the project root for full license information.
 //
 
+// TODO: The documentation says that System.Math is required for the graphics library buy is not referenced at this time. Should be reference it?
```
Comment (josesimoes) [inline nanoFramework.Graphics.Core/nanoFramework/UI/Mathematics.cs:7]: No need to. See previous comment on mscorlib.

**Review by josesimoes — COMMENTED** [review]
Just minor stuff.
Suggest that this added to mscorlib friends so that after the next release of mscorlib the Mathematics class can be removed.

**Comment by CoryCharlton** [comment]
I just replaced [Mathematics with MathInternal](https://github.com/nanoframework/nanoFramework.Graphics/pull/133/commits/ad51cc82e158ab518c191b16265f41dea5caa91f) so I think that covers all feedback in this PR.

File: nanoFramework.Graphics/nanoFramework.Graphics.nfproj [hunk 9]
```diff
@@ -277,9 +276,8 @@
     <None Include="packages.config" />
   </ItemGroup>
   <ItemGroup>
-    <Reference Include="mscorlib, Version=1.14.3.0, Culture=neutral, PublicKeyToken=c07d481e9758c731">
-      <HintPath>..\packages\nanoFramework.CoreLibrary.1.14.2\lib\mscorlib.dll</HintPath>
-      <Private>True</Private>
+    <Reference Include="mscorlib">
```
Comment (josesimoes) [inline nanoFramework.Graphics/nanoFramework.Graphics.nfproj:279]: I'm not very keen on having all these refs without the public key... 
Sometime ago I recall having to edit a bunch of project files manually to fix that.
Can you please give it a try and re-add these to fix this?
Are you using VS or what?

**Comment by josesimoes** [comment]
Merging as it is. The refs to mscorlib will be updated by automation after the release.

**Pull request merged**